lib/puppet/provider/exec/pwsh.rb

```diff
@@ -43,11 +43,28 @@ def validatecmd(command)
 
   private
 
+  # Returns the absolute path to pwsh using a cached result when possible
+  #
+  # @api private
+  # @return [String] the absolute path to the found pwsh executable.  Returns nil when it does not exist
+  def pwsh_command
+    return @pwsh_command if defined? @pwsh_command
+    @pwsh_command = get_pwsh_command
+  end
+
+  # Retrieves the absolute path to pwsh
+  #
+  # @api private
+  # @return [String] the absolute path to the found pwsh executable.  Returns nil when it does not exist
   def get_pwsh_command
     if @resource['path'].nil?
       pwsh = Puppet::Util.which('pwsh')
     else
-      pwsh = which_with_custom_env('pwsh', @resource['path'])
+      # Note that just like when we run the command in Puppet::Provider::Exec, the
+      # resource[:path] replaces the PATH, it doesn't add to it.
+      Puppet::Util.withenv({'PATH' => resource[:path].join(File::PATH_SEPARATOR)}, Puppet::Util::Platform.windows? ? :windows : :posix) do
+        pwsh = Puppet::Util.which('pwsh')
+      end
     end
     return pwsh unless pwsh.nil?
```

Comment:
So I think we can rework this method a bit to simplify / reduce the branching logic, which is a little difficult to follow.

```ruby
def get_pwsh_command
  search_paths = @resource['path'].nil? ? 
    ENV['PATH'] : 
    resource[:path].join(File::PATH_SEPARATOR)

  if Puppet::Util::Platform.windows?
    search_paths += ";#{ENV['ProgramFiles']}\\PowerShell\\6" + 
      ";#{ENV['ProgramFiles(x86)']}\\PowerShell\\6"
  end

  # Note that just like when we run the command in Puppet::Provider::Exec, the
  # resource[:path] replaces the PATH, it doesn't add to it.
  Puppet::Util.withenv({'PATH' => search_paths}, Puppet::Util::Platform.windows? ? :windows : :posix) do
    return Puppet::Util.which('pwsh')
  end
end
```

This disadvantage to this approach is that it's likely that `PATH` gets replaced with existing `PATH`, but given it simplifies the resolution logic, I would be inclined to lobby for it because it tidies up searching across the "well known" Windows install locations.

I'm also not clear why you note that `:path` attribute replaces `PATH`... but then still search these default locations on Windows?

Reply:
Yeah I was trying to avoid doing the PATH replacement if we didn't need it.  I do like the proposed change

Also there's no guarantee that the PS6 directories will be in the path, which is why I search the default locations if it couldn't be found in the PATH.

The `:path` comment is also a hint for future me, that we're not appending to the path like you would normally, but we're emulating the same behaviour that exec provider has, which is `use this path for searching`.

Reply:
Okay I've updated the search.  I used the UTF8 friendly `Puppet::Util.get_environment` and added some more comments e.g. case insensitive vars e.g. `Path` vs `PATH`

I also found `Puppet::Util.default_env` so cut down some code there too.